### PR TITLE
Import Excel datetimes with seconds

### PR DESCRIPTION
**Pull request by labkey-adam**
#### Rationale
We're failing to parse seconds from Excel datetime cells. https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=48930

We also want more consistent and obvious names for date-time formatting methods and constants.

#### Related Pull Requests
* https://github.com/LabKey/platform/pull/2942

### CODE REVIEW

File: api/src/org/labkey/api/reader/ExcelLoader.java [hunk 1]
```diff
@@ -970,7 +1001,8 @@ final Date getDateFromExcelDouble(double value)
 
         final String getIsoStringDateFromExcelDouble(double value)
         {
-            return DateUtil.formatDateTimeISO8601(getDateFromExcelDouble(value));
+            // Format as date time with seconds, Issue #48930
+            return DateUtil.formatDateTimeISO8601WithSeconds(getDateFromExcelDouble(value));
```
Comment (labkey-matthewb) [inline api/src/org/labkey/api/reader/ExcelLoader.java]: This is an odd function name.  ISO8601 includes seconds.  The version without seconds is just derived from LabKey's old default display format.

File: api/src/org/labkey/api/util/DateUtil.java [hunk 2]
```diff
@@ -1048,6 +1048,15 @@ public static String formatDateTimeISO8601(Date date)
...
     }
 
+    /**
+     * Format date and time using ISO 8601 pattern including seconds. This is appropriate only for persisting dates in
+     * machine-readable form, for example, for export or in filenames. Most callers should use
+     * formatDateTime(Container c, Date d) instead.
+     */
+    public static String formatDateTimeISO8601WithSeconds(Date date)
+    {
+        return formatDateTime(date, ISO_DATE_FORMAT_STRING + " " + LONG_TIME_FORMAT_STRING);
```
Comment (labkey-matthewb) [inline api/src/org/labkey/api/util/DateUtil.java]: As per note above LONG_TIME_FORMAT_STRING really should be called ISO_TIME_FORMAT_STRING.  What we call ISO_TIME_FORMAT_STRING isn't.

File: api/src/org/labkey/api/reader/ExcelLoader.java [hunk 3]
```diff
@@ -770,16 +773,44 @@ private static void checkData(ExcelLoader loader)
...
+                assertEquals(String.valueOf(Integer.MAX_VALUE), xl._toString(Integer.MAX_VALUE, true));
+                assertEquals("1.572", xl._toString("1.5720000000000001", true));
+                assertEquals("2.441", xl._toString("2.4409999999999998", true));
+                assertEquals("1.5720000000000001", xl._toString("1.5720000000000001", false));
+                assertEquals("2.4409999999999998", xl._toString("2.4409999999999998", false));
+                assertEquals("0.000001572", xl._toString("0.000001572", true));
+            }
+        }
+
+        @Test
```
Comment (labkey-matthewb) [inline api/src/org/labkey/api/reader/ExcelLoader.java:789]: Nice

**Review by labkey-matthewb — CHANGES_REQUESTED** [review]

File: api/src/org/labkey/api/reader/ExcelLoader.java (hunk 1 above)
Comment (labkey-susanh) [reply api/src/org/labkey/api/reader/ExcelLoader.java]: ISO8601 also apparently includes milliseconds. Do we abandon this name or actually want to use IS08601?

Comment (labkey-matthewb) [reply api/src/org/labkey/api/reader/ExcelLoader.java]: Yes, thanks for the correction.

Comment (labkey-adam) [reply api/src/org/labkey/api/reader/ExcelLoader.java]: My understanding is ISO 8601 is not a single format; the standard defines many, many date, time, and date+time formats. Formats can include space, T, Z, colons, etc. or not. Times are **not** required to have seconds, milliseconds, or even minutes. I believe all of the ISO constants are valid ISO formats and the ISO methods format using valid ISO formats.

Comment (labkey-adam) [reply api/src/org/labkey/api/reader/ExcelLoader.java]: I'm fine if we don't want to reference ISO 8601 and agree we should come up with a better naming system so we're clear which format is which being used. But let's not pretend there's only one true ISO format and the others are not ISO.

File: api/src/org/labkey/api/util/DateUtil.java (hunk 2 above)
Comment (labkey-adam) [reply api/src/org/labkey/api/util/DateUtil.java]: They're both ISO

**Review by labkey-matthewb — APPROVED** [review]

**Review by labkey-susanh — APPROVED** [review]

**Pull request merged**